src/DelegateRegistry.sol

```diff
@@ -104,199 +100,375 @@ contract DelegateRegistry is IDelegateRegistry {
+    function _computeDelegationHashForERC1155(address contract_, address delegate, bytes32 rights, uint256 tokenId, address vault)
+        private
+        pure
+        returns (bytes32)
     {
-        bytes32 delegationHash = _computeDelegationHashForERC1155(vault, delegate, contract_, tokenId, data);
-        // DONE (mireynolds): Why are we falling back to 721 checks instead of to contract checks?
-        // Was falling back to ERC721 in the case that ERC1155 tokenId is a non-fungible. Internally is unnecessary,
-        // so fallback now goes directly to contract, projects will just need to be aware that a 1155 nft delegation will
-        // return a balance rather than just the existence
-        return vaultDelegationHashes[vault].contains(delegationHash)
-            ? delegationInfo[delegationHash].balance
-            : (checkDelegateForContract(delegate, vault, contract_, data) ? type(uint256).max : 0);
+        return _encodeLastByteWithType(keccak256(abi.encode(contract_, delegate, rights, tokenId, vault)), DelegationType.ERC1155);
+    }
+
+    /// @dev Helper function that writes bytes32 data to delegation data location at array position
+    function _writeDelegation(bytes32 location, StoragePositions position, bytes32 data) private {
+        assembly {
+            sstore(add(location, position), data)
+        }
+    }
+
+    /// @dev Helper function that writes uint256 data to delegation data location at array position
+    function _writeDelegation(bytes32 location, StoragePositions position, uint256 data) private {
+        assembly {
+            sstore(add(location, position), data)
+        }
+    }
+
+    /// @dev Helper function that writes address data to delegation data location at array position
+    function _writeDelegation(bytes32 location, StoragePositions position, address data) private {
+        assembly {
+            sstore(add(location, position), data)
+        }
+    }
+
+    /// @dev Helper function to encode the last byte of a delegation hash to its type
+    function _encodeLastByteWithType(bytes32 _input, DelegationType _type) private pure returns (bytes32) {
+        return (_input & 0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF00) | bytes32(uint256(_type));
+    }
+
+    /// @dev Helper function to decode last byte of a delegation hash to obtain its type
+    function _decodeLastByteToType(bytes32 _input) private pure returns (DelegationType) {
+        return DelegationType(uint8(uint256(_input) & 0xFF));
+    }
+
+    /// @dev Helper function that filters an array of delegation hashes by removing disabled delegations and duplicates
+    function _filterDelegationHashes(bytes32[] memory array_) private view returns (bytes32[] memory) {
+        bool duplicate;
+        uint256 count = 0;
+        bytes32[] memory tempArray = new bytes32[](array_.length);
+
+        for (uint256 i = 0; i < array_.length; i++) {
```

Comment:
We get quadratic complexity in the length of the hashes here, could be a real problem for Liquid Delegates and other extensions as they scale, would be 500k operations at array length 1k and 12 million operations at array length 5k

Reply:
Latest push makes it impossible to push duplicates, and have removed the duplicates check from the delegations filter! :)